pallets/proofs-dealer/src/tests.rs

```diff
@@ -0,0 +1,347 @@
+        // Go past genesis block so events get deposited.
+        System::set_block_number(1);
+
+        // Create two users and add funds to the accounts.
+        let user_1 = RuntimeOrigin::signed(1);
+        let user_2 = RuntimeOrigin::signed(2);
+        let user_balance = 1_000_000_000_000_000;
+        assert_ok!(<Test as crate::Config>::NativeBalance::mint_into(
+            &1,
+            user_balance
+        ));
+        assert_ok!(<Test as crate::Config>::NativeBalance::mint_into(
+            &2,
+            user_balance
+        ));
+
+        // Mock two FileKeys.
+        let file_key_1 = BlakeTwo256::hash(b"file_key_1");
+        let file_key_2 = BlakeTwo256::hash(b"file_key_2");
+
+        // Dispatch challenge extrinsic twice.
+        assert_ok!(ProofsDealer::challenge(
+            RuntimeOrigin::signed(1),
+            file_key_1
+        ));
+
+        // Check that the event is emitted.
+        System::assert_last_event(
+            Event::NewChallenge {
+                who: 1,
+                file_key_challenged: file_key_1,
+            }
+            .into(),
+        );
+
+        assert_ok!(ProofsDealer::challenge(
+            RuntimeOrigin::signed(2),
+            file_key_2
+        ));
+
+        // Check that the event is emitted.
+        System::assert_last_event(
+            Event::NewChallenge {
+                who: 2,
+                file_key_challenged: file_key_2,
+            }
+            .into(),
+        );
+
+        // Check users' balance after challenge.
+        let challenge_fee: u128 = <Test as crate::Config>::ChallengesFee::get();
+        assert_eq!(
+            <Test as crate::Config>::NativeBalance::usable_balance(&1),
+            user_balance - challenge_fee
+        );
+        assert_eq!(
+            <Test as crate::Config>::NativeBalance::usable_balance(&2),
+            user_balance - challenge_fee
+        );
+
+        // Check that the challenge is in the queue.
+        let challenges_queue = crate::ChallengesQueue::<Test>::get();
+        assert_eq!(challenges_queue.len(), 2);
+        assert_eq!(challenges_queue[0], file_key_1);
+        assert_eq!(challenges_queue[1], file_key_2);
+    });
+}
+
+#[test]
+fn challenge_submit_existing_challenge_succeed() {
+    new_test_ext().execute_with(|| {
+        // Go past genesis block so events get deposited.
+        System::set_block_number(1);
+
+        // Create user and add funds to the account.
+        let user = RuntimeOrigin::signed(1);
+        let user_balance = 1_000_000_000_000_000;
+        assert_ok!(<Test as crate::Config>::NativeBalance::mint_into(
+            &1,
+            user_balance
+        ));
+
+        // Mock a FileKey.
+        let file_key = BlakeTwo256::hash(b"file_key");
+
+        // Dispatch challenge extrinsic twice.
+        assert_ok!(ProofsDealer::challenge(RuntimeOrigin::signed(1), file_key));
+        assert_ok!(ProofsDealer::challenge(RuntimeOrigin::signed(1), file_key));
```

Comment:
Should we allow multiple challenges for the same file_key? I only see this causing more transactions and work having to be done by storage providers.

Reply:
oh my bad misread.